Migrate to ACLINT

Mes0903 opened the pull request:
In this commit, based on the implementation of CLINT, an preliminary ACLINT is implemented, including the basic logic for operating `mtimer`, `mswi`, and `sswi`.  And this commit replaced CLINT with ACLINT, the old CLINT implementation was removed.

Currently, due to the lack of implementation, the introduced ACLINT uses only supervisor-level IPI. Therefore, although the logic for mswi is implemented, it is not being used at the moment.

It can be tested by `make check SMP=n`, where n is the number of harts you want to simulate.

chiangkd commented on aclint.c:
```diff
@@ -0,0 +1,220 @@
+{
+    if (!aclint_sswi_reg_write(sswi, addr, value << (RV_MEM_SW - width)))
+        vm_set_exception(hart, RV_EXC_STORE_FAULT, hart->exc_val);
+}
```
Let's do [add a new line to the end of a file](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file)

chiangkd commented on Makefile:
```diff
@@ -53,7 +53,7 @@ OBJS := \
 	uart.o \
 	main.o \
-	clint.o \
+	aclint.o \
```
I'm uncertain if the CLINT-related implementation will still exist once ACLINT is fully implemented. If it does, add an option here to use CLINT when ACLINT is unavailable.

Mes0903 commented on feature.h:
```diff
@@ -12,5 +12,10 @@
+/* ACLINT */
+#ifndef SEMU_FEATURE_ACLINT
+#define SEMU_FEATURE_ACLINT 1
+#endif
```
Added a flag `ENABLE_ACLINT` in the Makefile, allowing users to choose whether to enable ACLINT. By default, ACLINT is disabled (inspired by QEMU). If enabled, ACLINT will be used instead of CLINT.

Mes0903 commented:
In the latest commit, `ENABLE_ACLINT` was added to the Makefile. This flag determines whether to use CLINT or ACLINT, with CLINT being the default.

jserv commented:
> In the latest commit, `ENABLE_ACLINT` was added to the Makefile. This flag determines whether to use CLINT or ACLINT, with CLINT being the default. The ACLINT can be tested by `make check SMP=n ENABLE_ACLINT=1` now.

I don't think we should maintain both CLINT and ACLINT implementation.  The RISC-V Advanced Core Local Interruptor (ACLINT) enhances the existing SiFive CLINT design through several key improvements:
* Modular Architecture: ACLINT adopts a modular approach by separating timer and inter-processor interrupt (IPI) functionalities into distinct devices. This design enables RISC-V platforms to selectively implement only the necessary components, providing greater flexibility in system design.
* Dedicated S-Level IPI Device: The specification introduces a dedicated memory-mapped I/O (MMIO) device specifically for supervisor-level IPIs. This innovation eliminates the need to use Supervisor Binary Interface (SBI) calls for interrupt communication in Linux RISC-V systems, streamlining interrupt handling.
* Multi-Instance Support: ACLINT supports multiple timer and IPI device instances, a critical feature for multi-socket or multi-die Non-Uniform Memory Access (NUMA) system architectures. This capability allows for more complex and scalable system designs.
* Backward Compatibility: The specification maintains full compatibility with the original SiFive CLINT, ensuring that existing RISC-V platforms can seamlessly conform to the new ACLINT specification without requiring extensive redesigns or hardware modifications.

The recent Linux kernel now provides support for ACLINT, and QEMU has also integrated this feature, allowing users to enable it through the `-machine virt,aclint=on` option. Given these developments, it is an opportune time to transition to the ACLINT architecture.

Mes0903 commented:
The final commit removed all implementations of CLINT, including the previously introduced `ENABLE_ACLINT` and `OBJS_IR_CTRL`. Currently, the project only contains the implementation of ACLINT.

jserv commented:
Check https://cbea.ms/git-commit/ carefully and enforce the rules.

Mes0903 commented:
Due to two minor issues in the first commit of this PR—specifically, the absence of `#address-cells = <0>` causing DTC warnings, and the lack of a newline at the end of `aclint.c`—two additional small commits were added, making the commit history cluttered, also the commit message didn't follow the rule listed in https://cbea.ms/git-commit/.

To address this, the first commit message has been rewritten in accordance with the commit message rules, and the two minor issues have also been fixed in the process.

jserv reviewed (changes requested):
Don't use [backticks](https://en.wikipedia.org/wiki/Backtick) for the sake of terminal compatibility. Use single quotation marks instead.

jserv reviewed (changes requested):
Improve git commit messages by:
1. Clearly articulating the rationale behind transitioning from CLINT to ACLINT, highlighting the motivations and potential benefits of this migration.
2. Detailing the methodological approach for verifying and validating the functionalities of the proposed ACLINT support, including specific testing strategies or validation criteria.
3. Summarize the Linux kernel support on ACLINT along with git revisions.
4. Outlining potential future development paths to enhance and complete the SMP  support, identifying specific areas for improvement or expansion.

By the way, it is not necessary to mention "preliminary" if the proposed change covers the functionalities of CLINT. Instead, the major consideration is to migrate.

jserv commented on aclint.c:
```diff
@@ -0,0 +1,220 @@
+                                   uint32_t *value)
+{
+    /**
+     * @brief `addr & 0x4` is used to determine the upper or lower 32 bits
```
Don't use `@brief` notation. Instead, express in plain and informative English sentences.

jserv replied:
Don't use backtick. Use single quotation marks instead.

jserv commented on aclint.c:
```diff
@@ -0,0 +1,220 @@
+                                    uint32_t value)
+{
+    /**
+     * @brief The `cmp_val & 0xFFFFFFFF` is used to select the upper 32 bits
```
Ditto.

jserv commented on device.h:
```diff
@@ -172,27 +172,59 @@ void virtio_blk_write(hart_t *vm,
+
+/* ACLINT MSWI */
+typedef struct {
+    uint32_t msip[4096];  // Machine Software Interrupt Pending array
```
Always use C-style comments.

jserv commented on device.h:
```diff
@@ -172,27 +172,59 @@ void virtio_blk_write(hart_t *vm,
+
+/* ACLINT SSWI */
+typedef struct {
+    uint32_t ssip[4096];  // Supervisor Software Interrupt Pending array
```
Ditto. Explain why 4096 appears here.

Mes0903 commented:
The latest commit includes the following updates:

- Unified the use of C-style comments throughout the code.
- Removed the `@brief` notation, replacing it with plain and informative sentences.
- Replaced backticks with single quotation marks for consistency.
- Elaborated on the fundamental principles behind the transition from CLINT to ACLINT, emphasizing the motivations and benefits of this migration.
- Summarized Linux kernel support for ACLINT along with relevant git revisions.
- Provided detailed instructions on how to validate ACLINT functionalities, including specific testing strategies and validation methods.
- Outlined potential future development paths to enhance and complete SMP support, specifying areas for improvement or expansion.

jserv commented:
> The latest commit includes the following updates:

Don't attempt to summarize your changes. Instead, we always care about the changesets. Let's consolidate each git commit without unnecessary report.

jserv commented on aclint.c:
```diff
@@ -0,0 +1,218 @@
+    if (!aclint_mtimer_reg_read(mtimer, addr, value))
+        vm_set_exception(hart, RV_EXC_LOAD_FAULT, hart->exc_val);
+
+    *value >>= (RV_MEM_SW - width);
```
Rewrite as `*value >>= RV_MEM_SW - width` for short. The parentheses should only appear for the sake of the necessity.

jserv commented on aclint.c:
```diff
@@ -0,0 +1,218 @@
+    if (!aclint_mswi_reg_read(mswi, addr, value))
+        vm_set_exception(hart, RV_EXC_LOAD_FAULT, hart->exc_val);
+
+    *value >>= (RV_MEM_SW - width);
```
Ditto. Eliminate the parentheses.

jserv commented on device.h:
```diff
@@ -172,27 +172,95 @@ void virtio_blk_write(hart_t *vm,
+     * For more details, please refer to the register map at:
+     * https://github.com/riscv/riscv-aclint/blob/main/riscv-aclint.adoc#21-register-map
+     */
     uint64_t mtimecmp[4095];
```
The pre-allocated `mtimecmp` causes unexpected memory waste since we don't really need all of its members. Consider allocating it on-demand.

jserv commented on device.h:
```diff
@@ -172,27 +172,95 @@ void virtio_blk_write(hart_t *vm,
+     * For more details, please refer to the register map at:
+     * https://github.com/riscv/riscv-aclint/blob/main/riscv-aclint.adoc#31-register-map
+     */
+    uint32_t msip[4096];
```
Since the 4096th MSIP register is reserved, does it make sense to allocate 4096 elements?

Mes0903 replied:
ok, maybe I can allocate it based on the value of `SMP` specified during compilation.

jserv reviewed (changes requested):
Place proper `FIXME` and `TODO` comments to indicate the next step.

jserv commented on Makefile:
```diff
@@ -78,6 +78,7 @@ E :=
 S := $E $E
 
 SMP ?= 1
+CFLAGS += -DNUM_HARTS=$(SMP)
```
Don't do that. The use of ACLINT is transparent, meaning that it can be applied to uniprocessor configurations as well.

jserv commented on device.h:
```diff
@@ -172,6 +172,12 @@ void virtio_blk_write(hart_t *vm,
 
+/* ACLINT */
+
+#ifndef NUM_HARTS
+#define NUM_HARTS 1
+#endif
```
Ditto. Remove the uncertain code.

jserv commented on device.h:
```diff
@@ -190,7 +196,7 @@ typedef struct {
      * https://github.com/riscv/riscv-aclint/blob/main/riscv-aclint.adoc#21-register-map
      */
-    uint64_t mtimecmp[4095];
+    uint64_t mtimecmp[NUM_HARTS];
```
Allocate via runtime instead of build-time.

jserv commented on main.c:
```diff
@@ -302,7 +302,8 @@ static inline sbi_ret_t handle_sbi_ecall_IPI(hart_t *hart, int32_t fid)
 static inline sbi_ret_t handle_sbi_ecall_RFENCE(hart_t *hart, int32_t fid)
 {
-    /* TODO: RFENCE SBI extension */
+    /* TODO: Complete RFENCE SBI extension to support parallel multi-core
```
It was not meant to be "parallel." Check the spec.

Mes0903 replied:
Sorry for the imprecise wording. What I intended to express is that the current simulation executes sequentially. After reviewing the spec, I think the more precise term should be "concurrent multithreaded emulation".

jserv replied:
> Sorry for the imprecise wording. What I intended to express is that the current simulation executes sequentially. After reviewing the spec, I think the more precise term should be "concurrent multi-threaded emulation".

Use "multi-threaded system emulation" instead. See https://www.qemu.org/docs/master/devel/multi-thread-tcg.html
[Concurrency is not Parallelism](https://youtu.be/oV9rvDllKEg) by Rob Pike.

jserv replied:
> Sorry for the imprecise wording.

You don't have to say "sorry" early. The most polite way to express is to get things done right after you make a mistake.

jserv commented on main.c:
```diff
@@ -618,6 +619,7 @@ static int semu_start(int argc, char **argv)
     /* Emulate */
     uint32_t peripheral_update_ctr = 0;
     while (!emu.stopped) {
+        /* TODO: Implement parallel multi-core simulation */
```
Refine the wording.

Mes0903 commented:
Refine the wording in both comments and the previous git commit message. And use `calloc` to allocate the registers in ACLINT at runtime.

jserv commented:
Thank @Mes0903 for contributing!
